jetty-websocket/websocket-core/src/test/java/org/eclipse/jetty/websocket/testimpl/WebSocketHandler.java

```diff
@@ -0,0 +1,457 @@
+
+        @Override
+        public WebSocketCoreConnection newConnection(Connector connector, EndPoint endPoint, List<String> extensions)
+        {
+            ExtensionStack extensionStack = new ExtensionStack(extensionRegistry);
+            extensionStack.negotiate(objectFactory, policy, bufferPool,
+                                     extensions.stream().map(ExtensionConfig::parse).collect(Collectors.toList()));
+
+            UpgradeRequest upgradeRequest = null; // TODO
+            UpgradeResponse upgradeResponse = null; // TODO
+
+            return new WebSocketCoreConnection(
+                    endPoint,
+                    connector.getExecutor(),
+                    bufferPool,
+                    objectFactory,
+                    policy,
+                    extensionStack,
+                    upgradeRequest,
+                    upgradeResponse);
+        }
+    }
+
+
+    static class TestWebSocketSessionFactory extends ContainerLifeCycle implements WebSocketSessionFactory
+    {
+        @Override
+        public WebSocketCoreSession<ContainerLifeCycle,WebSocketCoreConnection,WebSocketLocalEndpoint,WebSocketRemoteEndpoint>
+        newSession(WebSocketCoreConnection connection, String subprotocol)
+        {
+
+            // TODO why is remoteEndpoint an interface rather than just an impl?
+            WebSocketRemoteEndpointImpl remoteEndpoint = new WebSocketRemoteEndpointImpl(connection);
+            remoteEndpoint.open(); // TODO why is this needed?
+
+            // TODO abstract the creation of a local Endpoint
+            WebSocketLocalEndpoint localEndpoint = new WebSocketLocalEndpoint()
+            {
+                @Override
+                public Logger getLog()
+                {
+                    return LOG;
+                }
+
+                @Override
+                public boolean isOpen()
+                {
+                    return connection.isOpen();
+                }
+
+                @Override
+                public void onOpen()
+                {
+                    LOG.debug("onOpen {}",this);
+                    TextFrame text = new TextFrame();
+                    text.setPayload("Opened!");
+
+                    // TODO do I need to mask this myself? if so then why?
+                    remoteEndpoint.sendFrame(text, new Callback()
+                    {
+                        @Override
+                        public void succeeded()
+                        {
+                            LOG.debug("onOpen write!");
+                        }
+
+                        @Override
+                        public void failed(Throwable x)
+                        {
+                            LOG.warn(x);
+                        }
+                    });
+                }
+
+                @Override
+                public void onText(Frame frame, Callback callback)
+                {
+                    ByteBuffer payload = frame.getPayload();
```

Comment:
Again, WebSocketLocalEndpoint is not an end user API.
You shouldn't be using it this way.

Reply:
I'm pretending to be then end user API.  So I'm doing what an end user API might do to get to the payload.     If I shouldn't be doing something, then the core API should not let me do it!